File: Data Structures & Algorithms/minimum-path-sum/submission-2.rs
impl Solution {

    pub fn min_path_sum(grid: Vec<Vec<i32>>) -> i32 {
        let (m, n) = (grid.len(), grid[0].len());
        let mut memo = vec![vec![-1; n];m];
        Self::dfs(&grid, 0, 0, &mut memo)
    }

    fn dfs(grid: &[Vec<i32>], row: usize, col: usize, memo: &mut Vec<Vec<i32>>) -> i32 {
        if row >= grid.len() || col >= grid[row].len() {
            return i32::MAX;
        }
        if row + 1 == grid.len() && col + 1 == grid[row].len() {
            return grid[row][col];
        }

        if memo[row][col] != -1 {
            return memo[row][col];
        }
        memo[row][col] = grid[row][col] + Self::dfs(grid, row + 1, col, memo).min(Self::dfs(grid, row, col + 1, memo));
        memo[row][col]
    }
    
}

File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-2.java
class Solution {
    int ROWS, COLS;
    int[][] memo;
    private final int[][] DIR = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    public int longestIncreasingPath(int[][] matrix) {
        ROWS = matrix.length;
        COLS = matrix[0].length;
        int len = 1;
        memo = new int[ROWS + 1][COLS + 1];
        for (int[] m: memo) {
            Arrays.fill(m, -1);
        }
        for (int row = 0; row < ROWS; ++row) {
            for (int col = 0; col < COLS; ++col) {
                len = Math.max(len, dfs(matrix, row, col));
            }
        }
        return len;
    }
    private int dfs(int[][] matrix, int row, int col) {

        
        if (memo[row][col] != -1) {
            return memo[row][col];
        }
        int len = 0;
        for (int[] dir: DIR ){
            int dr = dir[0], dc = dir[1];
            int newRow = row + dr, newCol = col + dc;
            if (newRow >= ROWS || newRow < 0 || newCol < 0 || newCol >= COLS || matrix[newRow][newCol] <= matrix[row][col]) {
                continue;
            }
            len = Math.max(len, dfs(matrix, newRow, newCol));
        }
        return memo[row][col] = 1 + len;
    }
}
